src/modules/user/services/user-wallet.service.ts

```diff
@@ -0,0 +1,104 @@
+import { Injectable, Inject } from "@nestjs/common";
+import { InjectRepository } from "@nestjs/typeorm";
+import { Repository } from "typeorm";
+import { utils } from "ethers";
+
+import { UserWallet } from "../model/user-wallet.entity";
+import { InvalidSignatureException, WalletNotFoundException } from "./exceptions";
+import { UserService } from "./user.service";
+
+@Injectable()
+export class UserWalletService {
+  constructor(
+    @InjectRepository(UserWallet) private userWalletRepository: Repository<UserWallet>,
+    @Inject(UserService) private userService: UserService,
+  ) {}
+
+  public async linkWallet({
+    userId,
+    walletAddress,
+    signature,
+    discordId,
+  }: {
+    userId: number;
+    walletAddress: string;
+    signature: string;
+    discordId: string;
+  }) {
+    this.verifySignedDiscordId({
```

Comment:
I think it's good to introduce the following assertion at the beginning of both POST and PATCH handler

```
1. select wallet with walletAddress
2. if wallet.userId === userId then do nothing and return data directly
3. if wallet.userId !== userId throw a BadRequest error
```

Reply:
done